### ISSUE TITLE
How to use tf.image.decodeImage with byteArray not with filePath In Java 0.3.1

### ISSUE THREAD

**Issue by mimilewis** [tensorflow/java#332]
We usually use tf.io.readFile(fileName) and then use tf.image.decodeImage, it  work nice in my code
now i want to decodeImage from raw byte[ ] ,so I change byte[ ] to String and use ：
       ` Graph g = new Graph();
            Ops tf = Ops.create(g);
Constant<TString> constant = tf.constant(new String(bytes));
DecodeJpeg.Options options = DecodeJpeg.channels(3L);
DecodeJpeg decodeImage = tf.image.decodeJpeg(constant, options);`
replace readFile.contents with constant

but it throw: Exception in thread "main" org.tensorflow.exceptions.TFInvalidArgumentException: Unknown image file format. One of JPEG, PNG, GIF, BMP required

I did a test that string value of byte[ ] and readFile was equals
so How to use tf.image.decodeImage with byteArray not with filePath In Java 0.3.1?

**Comment by pyrator** [comment]
You don't show very much of your code. The contents of readFile are a TString. Perhaps you should be converting your byte array to TString not String.

**Comment by mimilewis** [comment]
> You don't show very much of your code. The contents of readFile are a TString. Perhaps you should be converting your byte array to TString not String.

I didn't find a function to convert byte array TString. so I  use "String str = new String(bytes)" and then convert "str" to TSting with "tf.constant(str)".it doesn't work

**Comment by pyrator** [comment]
TString tString = TString.tensorOfBytes(NdArrays.scalarOfObject(bytes));
    Constant<TString> constant = tf.constant(tString);

**Comment by karllessard** [comment]
If you convert your bytes as a String and then pass that string to `tf.constant(str)` then you end up having a tensor of one single string value, which is not what you want here. You need to convert your array of bytes to a tensor of `TUInt8` where each value represents a R, G or B pixel value of your image.

The best way to do this is to wrap your standard Java array of bytes as a [`ByteDataBuffer`](https://github.com/tensorflow/java-ndarray/blob/main/ndarray/src/main/java/org/tensorflow/ndarray/buffer/ByteDataBuffer.java) and then call the `TUint8` constructor with that buffer and a given shape, which is normally (height, width, channels) or (channels, height, width depending in which format your image has been stored. And if your model has been trained with batches, don't forget to add this first dimension with `1` as a value.

In other words, your can do something like this:
```java
TUint8 image = TUint8.tensorOf(Shape.of(1, w, h, c), DataBuffers.of(bytes));
```

**Comment by karllessard** [comment]
(I've just deleted my previous solution since I've just realized that you are using `decodeJpeg` to read your image... sorry for that!)

Ok so for me, I guess the trick you initially tried (`Constant<TString> constant = tf.constant(new String(bytes))`) should have worked... Your image is really a Jpeg, right? If @pyrator solution above does not work, can you try this?

**Comment by karllessard** [comment]
Please see this example: https://github.com/tensorflow/java-models/blob/a9d975b0a0c2bcbe9a968e45f97ff54b98a83c32/tensorflow-examples/src/main/java/org/tensorflow/model/examples/cnn/fastrcnn/FasterRcnnInception.java#L251

**Comment by pyrator** [comment]
Just to confirm my solution should work if you are extracting your Jpeg from something like Blob storage or a Blob record in database, rather than converting a BufferedImage or OpenCV Mat to a byte array.

**Comment by mimilewis** [comment]
> ```
> TString tString = TString.tensorOfBytes(NdArrays.scalarOfObject(bytes));
> Constant<TString> constant = tf.constant(tString);
> ```


That's exactly what I want . Thanks for your help
